src/Codegen/InputObjectType.hack

```diff
@@ -0,0 +1,98 @@
+            ->setIsFinal()
+            ->setExtendsf('\%s', \Slack\GraphQL\Types\InputObjectType::class)
+            ->addTypeConstant(
+                $cg->codegenTypeConstant('TCoerced')->setValue('\\'.$hack_type, HackBuilderValues::literal()),
+            );
+
+        $name_const = $cg
+            ->codegenClassConstant('NAME')
+            ->setValue($this->input_type->getType(), HackBuilderValues::export());
+
+        $class->addConstant($name_const);
+
+        $ts = $this->reflection_type_alias->getResolvedTypeStructure();
+        invariant(
+            $ts['kind'] === \HH\TypeStructureKind::OF_SHAPE && !idx($ts, 'nullable', false),
+            'Input objects can only be generated from type aliases of a non-nullable shape type, got %s.',
+            TypeStructureKind::getNames()[$ts['kind']],
+        );
+
+        $class->addConstant(
+            $cg->codegenClassConstant('keyset<string> FIELD_NAMES')
+                ->setValue(Keyset\keys($ts['fields']), HackBuilderValues::export()),
+        );
+
+        // coerceFieldValues() and coerceFieldNodes()
+        $values = hb($cg)->addLine('return shape(')->indent();
+        $nodes = hb($cg)->addLine('return shape(')->indent();
+
+        foreach (($ts['fields'] as nonnull) as $field_name => $field_ts) {
+            $name_literal = \var_export($field_name, true);
+            $type = input_type(self::typeStructureToHackType($field_ts));
+            $values->addLinef('%s => %s->coerceValue($args[%s]),', $name_literal, $type, $name_literal);
+            $nodes->addLinef('%s => %s->coerceNode($args[%s], $vars),', $name_literal, $type, $name_literal);
+        }
+
+        $values->unindent()->addLine(');');
+        $nodes->unindent()->addLine(');');
+
+        return $class->addMethods(vec[
+            $cg->codegenMethod('coerceFieldValues')
+                ->setIsOverride()
+                ->addParameter('KeyedContainer<arraykey, mixed> $args')
+                ->setReturnType('this::TCoerced')
+                ->setBody($values->getCode()),
+            $cg->codegenMethod('coerceFieldNodes')
+                ->setIsOverride()
+                ->addParameterf('KeyedContainer<string, \\%s> $args', \Graphpinator\Parser\Value\Value::class)
+                ->addParameter('dict<string, mixed> $vars')
+                ->setReturnType('this::TCoerced')
+                ->setBody($nodes->getCode()),
+        ]);
+    }
+
+    private static function typeStructureToHackType<T>(TypeStructure<T> $ts): string {
+        $alias = Shapes::idx($ts, 'alias');
+        if ($alias is nonnull) {
+            return $alias;
+        }
+
+        switch ($ts['kind']) {
+            case TypeStructureKind::OF_INT:
+                return 'HH\int';
+            case TypeStructureKind::OF_STRING:
+                return 'HH\string';
+            case TypeStructureKind::OF_BOOL:
+                return 'HH\bool';
+            case TypeStructureKind::OF_VEC:
```

Comment:
Do we need another case here to handle nested shapes?

Reply:
Ah, I guess `$alias` takes care of that.

Reply:
Yeah the TypeStructure API is very confusing :( Various types get surfaced in very different ways. I'm honestly not sure if these cases cover everything, but one of the test input objects has a nested input objects so that part works...probably.